flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/TransformerBase.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.pipeline;
+
+import org.apache.flink.ml.api.core.Transformer;
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.ml.operator.batch.BatchOperator;
+import org.apache.flink.ml.operator.batch.source.TableSourceBatchOp;
+import org.apache.flink.ml.operator.stream.StreamOperator;
+import org.apache.flink.ml.operator.stream.source.TableSourceStreamOp;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.table.api.java.StreamTableEnvironment;
+
+/**
+ * The base class for transformer implementations.
+ *
+ * @param <T> A subclass of {@link TransformerBase}, used by {@link
+ *            org.apache.flink.ml.api.misc.param.WithParams}
+ */
+public abstract class TransformerBase<T extends TransformerBase<T>>
+	extends PipelineStageBase<T> implements Transformer<T> {
+
+	public TransformerBase() {
+		super();
+	}
+
+	public TransformerBase(Params params) {
+		super(params);
+	}
+
+	@Override
+	public Table transform(TableEnvironment tEnv, Table input) {
```

Comment:
Should we do a sanity check to make sure the tEnv passed in is the same table environment used by the input table?

Reply:
I agree with replace `void fit(TableEnvironment, Table)` method with the `MLEnvironmentFactory`, and I suggest we could discuss and change API in another PR.